pkg/logs/internal/launchers/file/provider/file_provider.go

```diff
@@ -0,0 +1,281 @@
+)
+
+// OpenFilesLimitWarningType is the key of the message generated when too many
+// files are tailed
+const openFilesLimitWarningType = "open_files_limit_warning"
+
+// wildcardOrdering controls what ordering is applied to wildcard files
+type wildcardOrdering string
+
+const (
+	// WildcardReverseLexicographical is the default option and does a pseudo reverse alpha sort
+	WildcardReverseLexicographical = "SortReverseLexicographical"
+	// WildcardMtime sorts based on the most recently modified time for each matching wildcard file
+	WildcardMtime = "SortMtime"
+)
+
+// selectionStrategy controls how the `filesLimit` slots we have are filled given a list of sources
+type selectionStrategy string
+
+const (
+	// GreedySelection will consider each source one-by-one, filling as many
+	// slots as is possible from that source before proceeding to the next one
+	GreedySelection = "ChooseGreedily"
+	// GlobalSelection will consider files from all sources together and will choose the
+	// top `filesLimit` files based on the `sortMode` ordering
+	GlobalSelection = "ChooseGlobally"
+)
+
+// FileProvider implements the logic to retrieve at most filesLimit Files defined in sources
+type FileProvider struct {
+	filesLimit      int
+	wildcardOrder   wildcardOrdering
+	selectionMode   selectionStrategy
+	shouldLogErrors bool
+}
+
+// NewFileProvider returns a new Provider
+func NewFileProvider(filesLimit int, wildcardOrder wildcardOrdering, selectionMode selectionStrategy) *FileProvider {
+	return &FileProvider{
+		filesLimit:      filesLimit,
+		wildcardOrder:   wildcardOrder,
+		selectionMode:   selectionMode,
+		shouldLogErrors: true,
+	}
+}
+
+type matchCount struct {
+	tracked int
+	total   int
+}
+
+// wildcardFileCounter tracks how many files a wildcard source matches, and how many of those are actually tailed.
+type wildcardFileCounter struct {
+	counts map[*sources.LogSource]matchCount
+}
+
+func newWildcardFileCounter() wildcardFileCounter {
+	return wildcardFileCounter{
+		counts: map[*sources.LogSource]matchCount{},
+	}
+}
+
+func (w *wildcardFileCounter) addTracked(src *sources.LogSource) {
+	matchCnt, _ := w.counts[src]
+	matchCnt.tracked++
+	w.counts[src] = matchCnt
+}
+
+func (w *wildcardFileCounter) setTotal(src *sources.LogSource, total int) {
+	matchCnt, _ := w.counts[src]
+	matchCnt.total = total
+	w.counts[src] = matchCnt
+}
+
+// FilesToTail returns all the Files matching paths in sources,
+// it cannot return more than filesLimit Files.
+// Files are collected according to the fileProvider's sortMode and chooseStrategy
+func (p *FileProvider) FilesToTail(inputSources []*sources.LogSource) []*tailer.File {
+	var filesToTail []*tailer.File
+	shouldLogErrors := p.shouldLogErrors
+	p.shouldLogErrors = false // Let's log errors on first run only
+
+	wildcardFileCounter := newWildcardFileCounter()
+	wildcardSources := make([]*sources.LogSource, 0, len(inputSources))
+
+	consumeFiles := func(files []*tailer.File) {
```

Comment:
Would it be possible to make this a regular method instead?

Reply:
I could make this a regular method instead, passing the appropriate vars explicitly instead of capturing them in the closure.
Does that option seem more readable to you?

Reply:
Yes, it would be more readable. It would also improve debugging: the function would have a proper name in tracebacks and profiles.